network/p2p/pinned_manager.go

```
@@ -0,0 +1,271 @@
+	}
+	return nil
+}
+
+// FilterUnpinned returns a new slice containing only peers that are not pinned.
+func (m *pinnedPeersManager) FilterUnpinned(ids []peer.ID) []peer.ID {
+	if len(ids) == 0 {
+		return ids
+	}
+	out := make([]peer.ID, 0, len(ids))
+	for _, id := range ids {
+		if !m.IsPinned(id) {
+			out = append(out, id)
+		}
+	}
+	return out
+}
+
+// ListPinned returns a copy snapshot.
+func (m *pinnedPeersManager) ListPinned() []peer.AddrInfo {
+	out := make([]peer.AddrInfo, 0)
+	m.peers.Range(func(_ peer.ID, ai *peer.AddrInfo) bool {
+		if ai != nil {
+			cp := peer.AddrInfo{ID: ai.ID, Addrs: append([]ma.Multiaddr(nil), ai.Addrs...)}
+			out = append(out, cp)
+		}
+		return true
+	})
+	// deterministic order by peer ID
+	slices.SortFunc(out, func(a, b peer.AddrInfo) int { return strings.Compare(a.ID.String(), b.ID.String()) })
+	return out
+}
+
+// PinPeer merges addresses, persists and protects; initiates a connect/backoff if necessary.
+func (m *pinnedPeersManager) PinPeer(ai peer.AddrInfo) error {
+	if ai.ID == "" {
+		return fmt.Errorf("empty peer id")
+	}
+	if len(ai.Addrs) == 0 {
+		return fmt.Errorf("pinned peer requires full multiaddr with address")
+	}
+	if m.host == nil {
+		return fmt.Errorf("p2p host is not initialized")
+	}
+	if existing, ok := m.peers.Get(ai.ID); ok && existing != nil {
+		existing = &peer.AddrInfo{ID: existing.ID, Addrs: p2pcommons.DedupMultiaddrs(append(existing.Addrs, ai.Addrs...))}
+		m.peers.Set(ai.ID, existing)
+	} else {
+		m.peers.Set(ai.ID, &peer.AddrInfo{ID: ai.ID, Addrs: p2pcommons.DedupMultiaddrs(append([]ma.Multiaddr(nil), ai.Addrs...))})
+	}
+	m.host.Peerstore().AddAddrs(ai.ID, ai.Addrs, peerstore.PermanentAddrTTL)
+	if m.connMgr != nil {
+		m.connMgr.Protect(ai.ID, m.protectTag)
+	}
+	go func(ai peer.AddrInfo) {
+		ctx, cancel := context.WithTimeout(m.ctx, 20*time.Second)
+		defer cancel()
+		_ = m.host.Connect(ctx, ai)
```

it would be good not to ignore the error here (I think I previously mentioned this in another review)